### PR TITLE
Avoid DQMIO file GUID collisions

### PR DESCRIPTION
**Pull request by dan131riley**
#### PR description:

As covered in #37240, the algorithm ROOT uses for `TUUID` can result in high filename collisions rates, which has been causing operational problems for the tier0.  Most framework uses of GUIDs were switched to using random GUIDs due to UUID collisions in containers.  This PR makes the same switch for DQMIO, and also restores the registration of DQMIO input files to the FJR.

#### PR validation:

Passes tests.  There's more validation I'd like to do, but I'm pushing this out now as there's high demand for a fix for this issue.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not a backport.  Backports to 12_3_0 and 12_2_0 will follow shortly.

### CODE REVIEW

**Comment by qliphy** [comment]
urgent

**Comment by qliphy** [comment]
please test

**Comment by qliphy** [comment]
type bugfix

File: DQMServices/FwkIO/plugins/DQMRootSource.cc [hunk 1]
```diff
@@ -533,7 +555,17 @@ std::shared_ptr<edm::FileBlock> DQMRootSource::readFile_() {
...
-    m_openFiles.insert(m_openFiles.begin(), file);
+    std::unique_ptr<std::string> guid{file->Get<std::string>(kCmsGuid)};
+
+    edm::JobReport::Token jrToken;
+    bool tokenValid{guid};
+
+    if (tokenValid) {
+      edm::Service<edm::JobReport> jr;
+      jrToken = jr->inputFileOpened(
+          pfn, lfn, std::string(), std::string(), "DQMRootSource", "source", *guid, std::vector<std::string>());
```
Comment (makortel) [inline DQMServices/FwkIO/plugins/DQMRootSource.cc:566]: I believe in principle we could report also the open/closes of the "old files" (that have UUID only via TFile), but I also suppose it wouldn't really matter in practice (assuming merging/harvesting old DQMIO files would be rare activity).

**Review by makortel — COMMENTED** [review]
Thanks Dan!

**Comment by jfernan2** [comment]
Upon implementing comments from @makortel , from DQM side, as far as I can tell with the unitTests available which I have made privately, the PR is fine. Thanks @dan131riley  !

File: DQMServices/FwkIO/plugins/DQMRootSource.cc [hunk 2]
```diff
@@ -308,6 +308,25 @@ struct DQMTTreeIO {
   };
 };
 
+struct openFileType {
```
[the remark this answers is not in the document]
Comment (dan131riley) [reply DQMServices/FwkIO/plugins/DQMRootSource.cc]: Moved to a nested type

File: DQMServices/FwkIO/plugins/DQMRootSource.cc (hunk 1 above)
Comment (dan131riley) [reply DQMServices/FwkIO/plugins/DQMRootSource.cc]: It also simplifies the code, cleaning up some parts that were a bit ugly, so I've implemented this and tested that it does the right thing with an input missing the CMS GUID

**Comment by dan131riley** [comment]
I've implemented Matti's suggestions and tested the handling of input files without the new GUID.

File: DQMServices/FwkIO/plugins/DQMRootSource.cc [hunk 3]
```diff
@@ -533,12 +551,20 @@ std::shared_ptr<edm::FileBlock> DQMRootSource::readFile_() {
     if (!isGoodFile && m_skipBadFiles)
       continue;
 
-    m_openFiles.insert(m_openFiles.begin(), file);
+    std::unique_ptr<std::string> guid{file->Get<std::string>(kCmsGuid)};
+    if (not guid) {
+      guid = std::make_unique<std::string>();
```
Comment (makortel) [inline DQMServices/FwkIO/plugins/DQMRootSource.cc:556]: I was actually thinking using the UUID of `TFile`, e.g. along
```suggestion
      guid = std::make_unique<std::string>(file->GetUUID().AsString());
      std::transform(guid->begin(), guid->end(), guid.begin(), (int (*)(int))std::toupper);
```
as it was before #28622.

Comment (dan131riley) [reply DQMServices/FwkIO/plugins/DQMRootSource.cc]: seems reasonable

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
ping bot

**Pull request merged**